lib/msf/core/handler/reverse_dns.rb

```diff
@@ -92,133 +93,178 @@ def start_handler
-        # to implement the Stream interface.
-        conn_threads << framework.threads.spawn("BindTcpHandlerSession", false, client) { |client_copy|
-          begin
-            client_copy.put([server_id.length].pack("C") + server_id)
-            handle_connection(wrap_aes_socket(client_copy), opts)
-          rescue
-            elog("Exception raised from BindTcp.handle_connection: #{$!}")
+      current_name = "STAGE"
+      loop do
+        begin          
+          session = nil
+          
+          #If last connection has a valid session or died        
+          if (framework.sessions.length > 0)
+            
+            framework.sessions.each_sorted do |k|
+              session = framework.sessions[k]
+            end 
+            current_name = session.machine_id.to_s
+          else
+            current_name = "STAGE"
+          end
+          
+          stime = Time.now.to_i
+          
+          if (current_name != "")       
+            
+            while (stime + ctimeout > Time.now.to_i)
+              begin
+                client = Rex::Socket::Tcp.create(
+                  'PeerHost' => rhost,
+                  'PeerPort' => lport.to_i,
+                  'Proxies'  => datastore['Proxies'],
+                  'Context'  =>
+                    {
+                      'Msf'        => framework,
+                      'MsfPayload' => self,
+                      'MsfExploit' => assoc_exploit
+                })
+              rescue Rex::ConnectionRefused
+                # Connection refused is a-okay
+                
+              rescue ::Exception
+                wlog("Exception caught in bind handler: #{$!.class} #{$!}")
+              end
+              
+              break if client
+              
+              # Wait a second before trying again
+              Rex::ThreadSafe.sleep(0.5)
+            end
+            
+            # Valid client connection?
+            if (client)
+              
+              # Increment the has connection counter
+              self.pending_connections += 1
+              
+              # Timeout and datastore options need to be passed through to the client
+              opts = {
+                :datastore    => datastore,
+                :expiration   => datastore['SessionExpirationTimeout'].to_i,
+                :comm_timeout => 60*60*24,
+                :retry_total  => datastore['SessionRetryTotal'].to_i,
+                :retry_wait   => datastore['SessionRetryWait'].to_i,
+                :timeout      => 60*20,
+                :send_keepalives => false
+              }
+              
+              
+              # Start a new thread and pass the client connection
+              # as the input and output pipe.  Client's are expected
+              # to implement the Stream interface.
+              conn_threads << framework.threads.spawn("BindDnsHandlerSession", false, client) { |client_copy|
```

Comment:
client_copy is a bit of a misnomer here - its the original object being passed into the thread. Example:
```
2.4.2 :001 > a = []; Thread.new(a) {|b| b << 1}
 => #<Thread:0x00000033659320b8@(irb):1 run> 
2.4.2 :002 > a
 => [1] 
```
If you dont want the original object, client.clone/dup/etc does the trick